File: tests/steps/checkout_steps.py
from pytest_bdd import given, then, when
from playwright.sync_api import expect


@given("I proceed to checkout")
def proceed_to_checkout(cart_page):
    cart_page.proceed_to_checkout()


@when("I enter checkout details:")
def enter_checkout_details(checkout_page, datatable):
    headers = [cell for cell in datatable[0]]
    values = [cell for cell in datatable[1]]
    row = dict(zip(headers, values))
    checkout_page.fill_customer_details(
        row["firstName"], row["lastName"], row["postalCode"]
    )


@when("I continue to the payment overview")
def continue_to_payment_overview(checkout_page):
    checkout_page.continue_to_payment()


@when("I continue to the payment overview without entering details")
def continue_without_details(checkout_page):
    checkout_page.continue_to_payment()


@when("I complete the order")
def complete_order(checkout_page):
    checkout_page.complete_order()


@then("I should see the order confirmation message")
def see_order_confirmation(checkout_page):
    checkout_page.expect_order_confirmation()


@then("I should see a checkout validation error")
def see_checkout_validation_error(page):
    expect(page.locator('[data-test="error"]')).to_be_visible()

File: tests/steps/checkout.steps.ts
import { createBdd } from 'playwright-bdd';
import { test, expect } from '../../fixtures/test-fixtures';

const { Given, When, Then } = createBdd(test);

Given('I proceed to checkout', async ({ cartPage }) => {
  await cartPage.proceedToCheckout();
});

When(
  'I enter checkout details:',
  async ({ checkoutPage }, dataTable) => {
    const row = dataTable.hashes()[0];
    await checkoutPage.fillCustomerDetails(row.firstName, row.lastName, row.postalCode);
  },
);

When('I continue to the payment overview', async ({ checkoutPage }) => {
  await checkoutPage.continueToPayment();
});

When('I continue to the payment overview without entering details', async ({ checkoutPage }) => {
  await checkoutPage.continueToPayment();
});

When('I complete the order', async ({ checkoutPage }) => {
  await checkoutPage.completeOrder();
});

Then('I should see the order confirmation message', async ({ checkoutPage }) => {
  await checkoutPage.expectOrderConfirmation();
});

Then('I should see a checkout validation error', async ({ page }) => {
  await expect(page.locator('[data-test="error"]')).toBeVisible();
});
